src/compiler/crystal/semantic/type_intersect.cr

```diff
@@ -82,7 +82,16 @@ module Crystal
       return nil if type1.instance_type.module?
 
       restricted = common_descendent(type1.instance_type, type2.instance_type.base_type)
-      restricted ? type1 : nil
+      restricted.try(&.virtual_type.metaclass)
```

Comment:
I changed this because restricting a type against a virtual type should always give a possibly virtual type: we don't know which of the actual subtypes it will be.